### DISCUSSION TITLE
I made a language server to support bibliographies in zk notes

### DISCUSSION THREAD

**Discussion by kha-dinh** [zk-org/zk#424] [Share]
I was annoyed by how incomplete the existing support for citation in my Neovim setup ([cmp-pandoc-references](https://github.com/jc-doyle/cmp-pandoc-references),  [mkdnflow](https://github.com/jakewvincent/mkdnflow.nvim)), so I went ahead and created a language server to support for bibtex-based citations. While [mkdnflow](https://github.com/jakewvincent/mkdnflow.nvim) does support similar features, my goal is to be editor-agnostic, more configurable,  and to fully leverage LSP capabilities (like `textDocument/hover`). Currently, the following capabilities are implemented:

| LSP Features                                                                                                                                    | Behavior                                                                                                                 |
| ----------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------ |
| [textDocument/definition](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition)  | Go to the first definition found in the `.bib` files.                                                                    |
| [textDocument/references](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_references)  | Find appearance of `prefix + ID` with ripgrep.                                                                           |
| [textDocument/hover](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover)            | Show metadata from `.bib` files based on configurations.                                                                 |
| [textDocument/completion](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion)  | Triggered by the `cite_prefix` configuration. Show completion of citation ID for Bibtex entries and their documentation. |
| [textDocument/diagnoistic](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_completion) | Find citations without a proper entry in the bibfile.               

Please check it out at [bibli-ls](https://github.com/kha-dinh/bibli-ls)! Any suggestions and ideas are welcome.

**Comment by tjex** [comment]
This is great. Thanks for sharing.
I take it the `.bib` file needs to be also stored at the root of the notebook / note repo?
I don't see any config options where a `.bib` file is pointed to.

**Reply by kha-dinh** [reply]
Currently it looks for a separated config file in the root of the project that points to the `.bib` file. There is a sample config in the README.
